### ISSUE TITLE
Profile vs subType vs domainType

### ISSUE THREAD

**Issue by neothemachine** [Reading-eScience-Centre/coveragejson#58]
Continuation of third bullet point of [comment in #56](https://github.com/Reading-eScience-Centre/coveragejson/issues/56#issuecomment-209825167):

> At the risk of opening a can of worms, I'm still not quite comfortable with the use of the word "profile" here. I can see the logic, but it could get confused with "vertical profile". Also, to my mind, a "profile" (in the sense of a specification) denotes a larger thing, e.g. a specialisation of a specification that is targeted to a particular community. I think in our cases these "profiles" are more like "feature types" or "subtypes".

It's not an easy one. But maybe you're right, since I had similar thoughts in [another issue](https://github.com/Reading-eScience-Centre/coveragejson/issues/50#issuecomment-206326842). The main reason why I found profiles to be a better match is that typically a type (sub type) provides some new structures or functionality, like in CIS's Grid vs point cloud which are two completely different things. Or like in GeoJSONs geometry types where the `"values"` array has a completely different structure in each type and you can't understand it without reading the type.
But in our case the domain structure itself already defines everything you need for every domain "type" and is not extended in any way, just restricted by how you can use it.

However, I agree that it is quite fine-grained in terms of profiles which are typically used for bigger things.

Quoting the RFC6906 again to have some context:
> A profile is defined not to alter the
   semantics of the resource representation itself, but to allow clients
   to learn about additional semantics (constraints, conventions,
   extensions) that are associated with the resource representation, in
   addition to those defined by the media type and possibly other
   mechanisms.

If we were to switch from "profile" to "subType", then it would be really just a renaming. There would still be a subType "GridCoverage" for a coverage, and a corresponding "Grid" subType for a domain. Is there anything else that would change? If not, what's the advantage really, except using a different name?

**Comment by jonblower** [comment]
I think I'm OK with "subtype", although I agree that this can imply that new structures are being introduced, whereas we are just introducing restrictions to help clients.

> what's the advantage really, except using a different name?

None really. I think "profile" is a technically accurate name, it's just that there's a potential confusion with "vertical profile".

Of course, we could change the name of the vertical profile instead! There may be advantages to doing this anyway, since "vertical profile" seems (from a Google search) to mean different things in different communities (e.g. sometimes it means the elevation profile of a hill along some transect). Maybe something to ask Unidata? VerticalColumn? Sounding? Hm.

**Comment by neothemachine** [comment]
(carrying over from https://github.com/Reading-eScience-Centre/coveragejson/issues/59#issuecomment-209947418)

> How might this work with the profiles that define specific domain types? Could a client say to the server "I only understand domain types that are defined in a certain profile?" So you could imagine a profile of CovJSON that defines typical coverages that are found in, say, hydrology, and another profile that defines coverages used in, say, climate science?
Does the Accept header allow you to specify multiple profiles?

I think you may be on to something. Yes, the profile field as defined in the RFC accepts multiple URIs seperated by whitespace, and those should then all be fulfilled.

To be honest, I haven't thought of creating a profile URI for the set of domain/coverage "types" that we currently define as "common profiles". But maybe this makes more sense, a bit like "I follow the CF-1.6 conventions", or in our case "I follow the common domain types specification 1.0".. or something like that. But then I'm not sure how to handle the domain subType field. If that is just a string like Grid, then that wouldn't have any meaning if you don't check for the profiles, meaning that every CovJSON document following those common domain types would have to include something like `"profiles": ["http://coveragejson.org/profiles/common-domain-types/1.0"]`. So that would mean the field "subType" is not even defined in the core CovJSON specification, but rather introduced in that common domain types profile, together with the names it defines. And this in turn means that someone else could invent his own domain type collection and use that instead: `"profiles": ["http://nasa.gov/covjson/better-domain-types/1.0"]`. I think that actually makes sense!

If you think about it, data publishers could even offer their data in multiple conventions if multiple turn out to be popular, and the client would know how to fetch it:

`Accept: application/prs.coverage+json; profile="http://coveragejson.org/profiles/common-domain-types/1.0"`

or

`Accept: application/prs.coverage+json; profile="http://nasa.gov/covjson/better-domain-types/1.0"`

Is that the solution?

**Comment by jonblower** [comment]
Just a few thoughts on this:

 1. Could the subType be a full URI (e.g. http://coveragejson.org/profiles/common/Grid) - the JSON-LD context could allow us to continue using the short form in the document itself.

 2. I don't see any harm in having a top-level `"profiles"` property that advertises any profiles. Is there less need for this if we use full URIs for subtypes?

 3. I don't quite see why `subType` would have to be defined in a profile. Why can't this be part of the core specification? The value of the subType property would be controlled by a profile, but not the property name itself, I think.

 4. There may be a case for saying that the "common domain types" profile is always explicitly included unless overridden in a particular document. It would be so common to use this that it might be tedious to specify it every time in the document and in Accept headers.

It may be worth thinking about what people might want to use profiles for. I can think of:
 1. Defining new domain types
 2. Restricting options to simplify clients, e.g. "Use UCUM only" to save clients having to understand all unit systems.
 3. "simple geospatial" in which CRS84 is the only permitted CRS.

**Comment by neothemachine** [comment]
To 1, yes, all concepts we define should be URIs and in the case of subType be a short form, doesn't matter if there's a JSON-LD context or not, this is always just bonus.

To 2, there is already a top-level `"profile"` property, so extending that to `"profiles"` as I suggested elsewhere probably makes sense. If you use full URIs for subtypes, well, then at least for that case you may not need a common-domain-types profile URL. I'm always a bit hesitant though with using full URIs everywhere for our own concepts since we don't have a stable namespace yet, and also it would make clients more annoying to write.

To 3, hm, I guess `subType` could be defined in the core specification. We already define `dataType` for axis objects which comes with 3 pre-defined ones (primitive, tuple, polygon) and importing a profile could import other dataTypes, similar to subType. Currently we force dataType to be a full URI if it's not a pre-defined one, which maybe is not a good idea. 

To 4, this is tricky, since it is not clear how to actually override a profile. And it is also dangerous since clients that don't understand the overriding profile would just ignore it and use the common domain types then, and if the overriding profile uses a different definition for what a Grid or Trajectory is then clients could misinterpret that or just stumble. I think therefore the common profile must be included if it is used. I can't see an easy way around that if we want to use it like that, but I do understand the redundancy for the 99% case. Maybe we have to think of something else.

A note about profiles and Accept headers / content negotiation. Realistically I think that data providers will not offer different variants of coverages and then offer them like that by handling profiles in Accept headers. At least in the near future. So, it is a bit idealistic, but of course that shouldn't prevent us from building in the option.

**Comment by neothemachine** [comment]
Notes of offline discussion today:
- Profiles as a concept are probably too generic in the context of covjson, at least for domain/coverage (sub)types.
- The common domain/coverage subtypes that we define should be part of the main specification (or as it is now, linked from that specification, even if they are two separate documents), and if people want to use custom types they should use full URIs to avoid conflicts if such a custom type makes it into the covjson specification, in possibly adapted form.
- Instead of "profile", it would be "subType".
- Whether "profile" / "profiles" stay is another question, but at least for now there doesn't seem to be a compelling use case. The media type still includes the "profile" parameter though and there is a single profile which is the stand-alone profile which may be used in content negotiation to request a stand-alone covjson document (embedded domain & range). The stand-alone profile however wouldn't be included in the covjson document itself since the fact that it is stand-alone is very easy to check and including the profile seems redundant.

Did I miss anything?

**Comment by jonblower** [comment]
I think you’ve captured everything. Is it worth considering alternative words for “subType”, if you feel that implies something too specific? “domainType” perhaps?

**Comment by jonblower** [comment]
Similarly we could consider "coverageType" at the coverage level.

**Comment by neothemachine** [comment]
Possibly:
```js
{
  "type": "Coverage",
  "coverageType": "GridCoverage"
  "domain": {
    "type": "Domain",
    "domainType": "Grid",
    ...
  }
}
```
I think this doesn't win anything though over subType in terms of what it implies, except the fact that it disambiguates a bit and that you can copy "domainType" as-is to the root coverage object if the domain is not embedded. Btw, we appear two times on google first page for "coverage domain type". Not sure if that's good or bad!

**Comment by jonblower** [comment]
I think if there's no loss I would prefer `domainType` and `coverageType` for the reasons you mention, also it's just a little clearer. We can also use the phrase "Domain types" and "Coverage types" in documents and then the headings in the documents will match nicely with the terms in the format.

> Btw, we appear two times on google first page for "coverage domain type"
Only once for me (the blog entry from the MELODIES website)! Probably depends on your search history...

**Comment by neothemachine** [comment]
I have to admit I'm not comfortable yet with `domainType` and `coverageType`, especially the former.

If you look at that structure:
```js
{
  "type": "Coverage",
  "domain": {
    "type": "Domain",
    "domainType": "Grid",
    ...
  },
  ...
}
```
then it just feels weird reading the JSON: "There's an object of type Coverage, then that has a domain which is an object of type Domain. And the type of the domain object is Grid?" When I see the structure above then it looks like the type of the associated domain is "Domain", which then is confusing by having an extra domain type property. On its own it makes kind of sense, but in this Coverage object structure it feels a bit weird.

GeoJSON solved this by not having a root type `Geometry` but directly using the sub types as `"type": "Polygon"`. Of course, that only works because they have a limited number of geometry types (=simple geometries), which is not the case for us.

From that point of view I would find `subType` better (and then `domainSubType` in the coverage object when the domain is not embedded), but it is uglier.

Is there another word that doesn't have "type" in it? I still think that "type" implies too much and it doesn't sound like something optional.

**Comment by jonblower** [comment]
I see what you mean. I'm also happy with `subType` / `domainSubType`, although this isn't perfect either. I can't think of a better alternative, so either way we need to explain the meaning of these terms in the docs. Maybe the community can think of something better.

But just to explore all the options - would it not be OK to do something like:

```js
{
  "type": "GridCoverage",
  "domain": {
    "type": "GridDomain",
   ...
  }
}
```

So we define a load of subtypes of Coverage and Domain in some kind of vocabulary somewhere, which  RDF/JSON-LD clients could understand. For non-RDF clients, we have the simple rule that all types must end in the string "Coverage" or "Domain", so they can identify object types correctly.

**Comment by neothemachine** [comment]
I'm really not a fan of `"type": "GridCoverage"`. It merges two aspects and gives even more meaning to "Grid". And it just smells wrong if clients would have to do string suffix parsing just to find out that something is a coverage or domain. Seems like overcompacting things at the expense of client complication.

**Comment by jonblower** [comment]
Maybe `subType` is the best (least worst) option for now then?

**Comment by neothemachine** [comment]
There's still `profile` though, which is getting a bit more attractive again I think.

**Comment by jonblower** [comment]
I must admit I still prefer `subtype` to `profile`. A GridCoverage *is* a subtype of a Coverage, is it not? At least at the conceptual level. This feels more accurate to me than saying that it's a profile of a coverage, also not forgetting the name clash with VerticalProfile and potential other uses of "profile" (e.g. the standalone profile).

**Comment by neothemachine** [comment]
Ok... final attempt. How about we use `domainType` and get rid of `coverageType` and the same for collection. Then the idea is that you can use `domainType` at all object hierarchy levels and it means that it applies to all elements underneath it. For example, if a coverage collection has `"domainType": "Grid"` then it means that all coverages in it have the same, and this means that all domains have the same. It gets inherited. I don't see a need for `coverageType` since this just defines the domain type and doesn't put any constraints on the rest of the coverage (parameters, range), and having two different concepts (or three, if collections also have that) means more code to implement.

I [wrote earlier](https://github.com/Reading-eScience-Centre/coveragejson/issues/58#issuecomment-215367118) that I didn't like `domainType` as it may be confusing when reading the object structure, but I think it is not that bad in the end, there doesn't seem to be a perfect solution anyway.

So by restricting this whole profile/subType.. business to domain types we can concentrate on that aspect only and don't have to think about other implications/constraints that possibly someone may define at coverage/collection level which are not related to domain types. I think by doing that we define the most useful thing that we currently need and leave the door open to extend those ideas in the future if needed (by adding additional properties).

What do you think?

**Comment by jonblower** [comment]
I like this idea. Nice and simple, fairly intuitive and neatly encapsulates the idea that all coverages are basically the same, differing only in their domain type.
